core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1324,9 +1324,16 @@ public final void messageReceived(BytesReference reference, Channel channel, Str
                 }
                 streamIn = compressor.streamInput(streamIn);
             }
-            if (version.isCompatible(getCurrentVersion()) == false) {
+            Version compatibilityVersion = TransportStatus.isHandshake(status) ?
+                // for handshakes we are compatible with N-2 since otherwise we can't figure out our initial version
```

Comment:
I think this'd be more readable if the comment was above the variable. I don't know enough about the handshake to verify the change off hand though.